File: DocumentAnalysisSystem/Visitors/KeywordSearchVisitor.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using DocumentAnalysisSystem.Models;

namespace DocumentAnalysisSystem.Visitors
{
    public class KeywordSearchVisitor : IDocumentVisitor
    {
        public KeywordSearchVisitor(string keyword)
        {
            Keyword = keyword.Trim();
        }

        public string Keyword { get; }
        public int TotalMatches { get; private set; }
        public IReadOnlyList<string> KeywordDetails => _keywordDetails.AsReadOnly();
        private readonly List<string> _keywordDetails = new();

        public void Visit(Document document)
        {
            // No action needed at document level.
        }

        public void Visit(Section section)
        {
            // No direct search inside section nodes.
        }

        public void Visit(Paragraph paragraph)
        {
            if (string.IsNullOrWhiteSpace(Keyword) || string.IsNullOrWhiteSpace(paragraph.Text))
                return;

            var pattern = $"\\b{Regex.Escape(Keyword)}\\b";
            var matches = Regex.Matches(paragraph.Text, pattern, RegexOptions.IgnoreCase);
            if (matches.Count > 0)
            {
                TotalMatches += matches.Count;
                _keywordDetails.Add($"{matches.Count} lần trong đoạn: {paragraph.Text.Trim()}");
            }
        }
    }
}

File: DocumentAnalysisSystemJava/src/main/java/com/documentanalysis/visitors/KeywordSearchVisitor.java
package com.documentanalysis.visitors;

import com.documentanalysis.models.Document;
import com.documentanalysis.models.Paragraph;
import com.documentanalysis.models.Section;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class KeywordSearchVisitor implements IDocumentVisitor {
    private final String keyword;
    private final List<String> keywordDetails = new ArrayList<>();
    private int totalMatches;

    public KeywordSearchVisitor(String keyword) {
        this.keyword = keyword == null ? "" : keyword.trim();
    }

    public String getKeyword() {
        return keyword;
    }

    public int getTotalMatches() {
        return totalMatches;
    }

    public List<String> getKeywordDetails() {
        return Collections.unmodifiableList(keywordDetails);
    }

    @Override
    public void visit(Document document) {
        // No direct search at document level.
    }

    @Override
    public void visit(Section section) {
        // No direct search at section level.
    }

    @Override
    public void visit(Paragraph paragraph) {
        if (keyword.isBlank() || paragraph.getText() == null || paragraph.getText().isBlank()) {
            return;
        }

        Pattern pattern = Pattern.compile("\\b" + Pattern.quote(keyword) + "\\b", Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(paragraph.getText());
        int matches = 0;
        while (matcher.find()) {
            matches++;
        }

        if (matches > 0) {
            totalMatches += matches;
            keywordDetails.add(matches + " lần trong đoạn: " + paragraph.getText().trim());
        }
    }
}
